File: lab2_ws/src/safety_node/scripts/safety_node.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

import numpy as np
from sensor_msgs.msg import LaserScan
from nav_msgs.msg import Odometry
from ackermann_msgs.msg import AckermannDriveStamped, AckermannDrive


class SafetyNode(Node):
    """
    The class that handles emergency braking.
    """

    def __init__(self):
        super().__init__('safety_node')
        """
        One publisher should publish to the /drive topic with a AckermannDriveStamped drive message.

        You should also subscribe to the /scan topic to get the LaserScan messages and
        the /ego_racecar/odom topic to get the current speed of the vehicle.

        The subscribers should use the provided odom_callback and scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        """
        self.speed = 0.0

        self.declare_parameter('ttc_tol', 1.5)
        self.scan_sub_ = self.create_subscription(LaserScan, '/scan', self.scan_callback, 10)
        self.odom_sub_ = self.create_subscription(Odometry, '/ego_racecar/odom', self.odom_callback, 10)
        self.publisher_ = self.create_publisher(AckermannDriveStamped, '/drive', 10)

    def odom_callback(self, odom_msg: Odometry):
        self.speed = odom_msg.twist.twist.linear.x

    def scan_callback(self, scan_msg: LaserScan):
        n_ranges = len(scan_msg.ranges)
        ranges = np.array(scan_msg.ranges)
        in_bound = (ranges >= scan_msg.range_min) & (ranges <= scan_msg.range_max)
        ranges = np.where(in_bound, ranges, np.inf)
        angles = np.arange(n_ranges) * scan_msg.angle_increment + scan_msg.angle_min
        vels = self.speed * np.cos(angles)
        vels = np.where(vels > 0, vels, 1e-8)
        ittc = min(ranges / vels)
        self.get_logger().info("iTTC: %f" % ittc)
        if ittc < self.get_parameter('ttc_tol').get_parameter_value().double_value:
            message = AckermannDriveStamped()
            message.drive.speed = 0.0
            self.publisher_.publish(message)


def main(args=None):
    rclpy.init(args=args)
    safety_node = SafetyNode()
    rclpy.spin(safety_node)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    safety_node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: lab2_ws/src/safety_node/src/safety_node.cpp
#include "rclcpp/rclcpp.hpp"
/// CHECK: include needed ROS msg type headers and libraries
#include "sensor_msgs/msg/laser_scan.hpp"
#include "nav_msgs/msg/odometry.hpp"
#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"

using std::placeholders::_1;

class Safety : public rclcpp::Node {
// The class that handles emergency braking

public:
    Safety() : Node("safety_node") {
        /*
        You should also subscribe to the /scan topic to get the
        sensor_msgs/LaserScan messages and the /ego_racecar/odom topic to get
        the nav_msgs/Odometry messages

        The subscribers should use the provided odom_callback and 
        scan_callback as callback methods

        NOTE that the x component of the linear velocity in odom is the speed
        */

        this->declare_parameter("ttc_tol", 1.5);

        scan_sub_ = this->create_subscription<sensor_msgs::msg::LaserScan>(
                "/scan", 10, std::bind(&Safety::scan_callback, this, _1));
        odom_sub_ = this->create_subscription<nav_msgs::msg::Odometry>(
                "/ego_racecar/odom", 10, std::bind(&Safety::drive_callback, this, _1));
        publisher_ = this->create_publisher<ackermann_msgs::msg::AckermannDriveStamped>(
                "/drive", 10);
    }

private:
    double speed = 0.0;

    rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr scan_sub_;
    rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr odom_sub_;
    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr publisher_;

    void drive_callback(const nav_msgs::msg::Odometry::ConstSharedPtr msg) {
        speed = msg->twist.twist.linear.x;
    }

    void scan_callback(const sensor_msgs::msg::LaserScan::ConstSharedPtr scan_msg) {
        float iTTC = 1e8;
        for (int i = 0; i < (int) scan_msg->ranges.size(); ++i) {
            float curr_range = scan_msg->ranges[i];
            if (curr_range > scan_msg->range_max || curr_range < scan_msg->range_min) {
                continue;
            }
            float curr_angle = scan_msg->angle_min + scan_msg->angle_increment * float(i);
            float curr_vel = (float) speed * std::cos(curr_angle);
            if (curr_vel <= 0) {
                curr_vel = 1e-8;
            }
            float curr_ittc = curr_range / curr_vel;
            iTTC = std::min(iTTC, curr_ittc);
        }
        RCLCPP_INFO(this->get_logger(), "iTTC: %f", iTTC);
        if (iTTC < this->get_parameter("ttc_tol").as_double()) {
            auto message = ackermann_msgs::msg::AckermannDriveStamped();
            message.drive.speed = 0.0;
            publisher_->publish(message);
        }
    }
};

int main(int argc, char **argv) {
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Safety>());
    rclcpp::shutdown();
    return 0;
}